### PR TITLE
[Aptos Framework] [Account] support multi_ed25519 key rotation

### PR DESCRIPTION
**Pull request by chloeqjz**
### Description
Support multi_ed25519 key rotation via https://github.com/aptos-labs/aptos-core/pull/3552

### Test Plan
e2e tests for rotating from ed25519 key to ed25519 key or multi_ed25519 key
move unit tests for rotating from multi_ed25510 key to ed25519 key or multi_ed25519 key

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3626)

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 1]
```diff
@@ -249,6 +254,81 @@ module aptos_framework::account {
         account_resource.authentication_key = new_auth_key;
     }
 
+    fun verify_key_rotation_signature(scheme: vector<u8>, public_key_bytes: vector<u8>, signature: vector<u8>, challenge: RotationProofChallenge) : vector<u8> {
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:257]: scheme should just be u8

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 2]
```diff
@@ -249,6 +254,81 @@ module aptos_framework::account {
...
+
+        challenge
+    }
+
+    public entry fun rotate_authentication_key(
+        account: &signer,
+        in_scheme: vector<u8>,
+        to_scheme: vector<u8>,
+        from_public_key_bytes: vector<u8>,
+        to_public_key_bytes: vector<u8>,
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:303]: minor nit... order from_scheme, from_bytes, to_scheme, to_bytes
and keep consistent naming
that will make it easier for most users since names and ordering map flow

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 3]
```diff
@@ -249,6 +254,81 @@ module aptos_framework::account {
...
+    }
+
+    fun get_challenge_message(account_address: address, in_scheme: vector<u8>, from_public_key_bytes: vector<u8>, to_public_key_bytes: vector<u8>): RotationProofChallenge acquires Account {
+        let account_resource = borrow_global_mut<Account>(account_address);
+
+        if (in_scheme == ED25519_SCHEME) {
+            let from_pk = ed25519::new_unvalidated_public_key_from_bytes(from_public_key_bytes);
+            let from_auth_key = ed25519::unvalidated_public_key_to_authentication_key(&from_pk);
+            assert!(account_resource.authentication_key == from_auth_key, error::unauthenticated(EWRONG_CURRENT_PUBLIC_KEY));
+        } else {
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:280]: should have else if and finally just a fail / panic in the else case for unsupported

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 4]
```diff
@@ -249,6 +254,81 @@ module aptos_framework::account {
...
+            let pk = multi_ed25519::new_unvalidated_public_key_from_bytes(public_key_bytes);
+            let sig = multi_ed25519::new_signature_from_bytes(signature);
+            assert!(multi_ed25519::signature_verify_strict_t(&sig, &pk, challenge), std::error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+            auth_key = multi_ed25519::unvalidated_public_key_to_authentication_key(&pk);
+        };
+        auth_key
+    }
+
+    fun get_challenge_message(account_address: address, in_scheme: vector<u8>, from_public_key_bytes: vector<u8>, to_public_key_bytes: vector<u8>): RotationProofChallenge acquires Account {
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:273]: might not make sense to have a dedicated function without it being called else where --- it isn't that much code.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 5]
```diff
@@ -249,6 +254,81 @@ module aptos_framework::account {
...
+        let curr_auth_key = from_bcs::to_address(account_resource.authentication_key);
+
+        let challenge = RotationProofChallenge {
+            sequence_number: account_resource.sequence_number,
+            originator: account_address,
+            current_auth_key: curr_auth_key,
+            new_public_key: to_public_key_bytes,
+        };
+
+        challenge
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:295]:
```suggestion
        RotationProofChallenge {
            sequence_number: account_resource.sequence_number,
            originator: account_address,
            current_auth_key: curr_auth_key,
            new_public_key: to_public_key_bytes,
        }
```

**Review by davidiw — COMMENTED** [review]
will review more tomorrow, but this is ... amazing

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 1 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:257]: Also, can you please take the `rotation_challenge` by reference, to avoid an unnecessary copy when you call this function twice in `rotate_authentication_key`?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 4 above)
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:273]: The name `get_challenge_message` is not sufficiently informative here because this function also checks that the input "from PK" is indeed the current PK set at the input account address.

This needs to be renamed, or this functionality should be split out into a different function.

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 3 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:280]: If so, the same should be done for `verify_key_rotation_signature`

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 4 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:273]: I would opt for separating the functionality.

Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:273]: Or opt for this!

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 2 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:303]: Yes, can we use consistent naming: e.g., use "from_scheme" (rather than "in_scheme")

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 1 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move]: yep just updated! thanks for the review : )

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 3 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:294]: oh I was thinking line 284-285 should have already covered the case: 
` assert!((from_scheme == ED25519_SCHEME || from_scheme == MULTI_ED25519_SCHEME), EINVALID_SCHEME);
   assert!((to_scheme == ED25519_SCHEME || to_scheme == MULTI_ED25519_SCHEME), EINVALID_SCHEME);
`

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 6]
```diff
@@ -249,6 +254,77 @@ module aptos_framework::account {
...
+            new_public_key: to_public_key_bytes,
+        };
+
+        // verify the signatures from the current private key and the new private key are valid
+        let curr_auth_key = verify_key_rotation_signature(from_scheme, from_public_key_bytes, cap_rotate_key, &challenge);
+        let new_auth_key = verify_key_rotation_signature(to_scheme, to_public_key_bytes, cap_update_table, &challenge);
+
+        // update the address_map table, so that we can reference to the originating address using the current address
+        let address_map = &mut borrow_global_mut<OriginatingAddress>(@aptos_framework).address_map;
+        let originating_address = signer::address_of(account);
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:314]: You've already computed this above in addr on line 282.

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 4 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move]: okie just merged this with the rotate key function!

**Comment by boczeratul** [comment]
Kudos to @chloeqjz and @davidiw. Really appreciate the work here. Saved our day 🎊

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 7]
```diff
@@ -249,6 +254,77 @@ module aptos_framework::account {
...
+        let curr_auth_key = verify_key_rotation_signature(from_scheme, from_public_key_bytes, cap_rotate_key, &challenge);
+        let new_auth_key = verify_key_rotation_signature(to_scheme, to_public_key_bytes, cap_update_table, &challenge);
+
+        // update the address_map table, so that we can reference to the originating address using the current address
+        let address_map = &mut borrow_global_mut<OriginatingAddress>(@aptos_framework).address_map;
+        let originating_address = addr;
+        let curr_address = from_bcs::to_address(curr_auth_key);
+        let new_address = from_bcs::to_address(new_auth_key);
+        if (table::contains(address_map, curr_address)) {
+            originating_address = table::remove(address_map, curr_address);
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:318]: I'm a bit confused by this: shouldn't this be an `assert!(originating_address == table::remove(address_map, curr_address), 1);`? 

In other words, this assigns the same value to `originating_address`, no?

In other _other_ words, whatever originating address we remove from table[curr_auth_key], should be the original address of the signer computed on line 282 at the beginning of `rotate_authentication_key`, no? After all, the signer's account has `curr_auth_key` set as its current authentication key (as per the check on line 293).

Maybe I'm missing something / too tired.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 8]
```diff
@@ -518,4 +592,48 @@ module aptos_framework::account {
...
+
+        let curr_pk_with_scheme = copy curr_pk_bytes;
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:604]: Can we use the API from (Multi)Ed25519 for this instead of doing it manually??

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 9]
```diff
@@ -249,6 +254,77 @@ module aptos_framework::account {
...
+
+    public entry fun rotate_authentication_key(
+        account: &signer,
+        from_scheme: u8,
+        from_public_key_bytes: vector<u8>,
+        to_scheme: u8,
+        to_public_key_bytes: vector<u8>,
+        cap_rotate_key: vector<u8>,
+        cap_update_table: vector<u8>,
+    ) acquires Account, OriginatingAddress {
```
Comment (jjleng) [inline aptos-move/framework/aptos-framework/sources/account.move:281]: nit: this needs some documentation

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 3 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:294]: Yes, indeed... Might be overkill.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 10]
```diff
@@ -518,4 +592,48 @@ module aptos_framework::account {
...
+        let new_pk_bytes = x"20fdbac9b10b7587bba7b5bc163bce69e796d71e4ed44c10fcb4488689f7a144";
+        let cap_rotate_key = x"0bc503a99ee09a2bfaeb0039a092abda54cf7493608c01a2e0ac4a0c49958fcbf7eb0521e388ec73b03b978dce79ffda20194aca52cdd13f35c4776de8d27808f0d8c0dbeb14700b46e3c927d848aeba74e0749cdc6429fa1aba1d3e7ef57948bef0810125ccaa2de25a167d13f5725bbc85fcac1b03dff944275d4b4cad3c0ac0000000";
+        let cap_update_table = x"dcb63645f22c9c3f9ff6b05293dc3c0e22e4bd6d6c4001d68869139e78a645d4c0745b61538916b0f6e42736f0dbba19dbd6d1eee5bdd5ef3e7c1d0617b72d01";
+
+        let curr_pk_with_scheme = copy curr_pk_bytes;
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:626]: Same request here to use the auth key derivation API!

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 9 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:289]: just added some documentation! thanks for reviewing : )

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 8 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move]: great point! just updated

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move]: ah yea I think you're right. 

I was thinking of the usecase where a user rotates from `key_a` to `key_b`, and now they want to rotate to `key_c`. if they call the function from `account_b` - it's not the originating account so the address of that account is not the `originating_address`, and in that case the `originating_address` would be the value we removed from the table? but I think in general, we're expecting the user to call this function from `account_a`. 

just updated this part of code.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 11]
```diff
@@ -249,6 +255,84 @@ module aptos_framework::account {
         account_resource.authentication_key = new_auth_key;
     }
 
+    fun verify_key_rotation_signature(scheme: u8, public_key_bytes: vector<u8>, signature: vector<u8>, challenge: &RotationProofChallenge) : vector<u8> {
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:258]: should we call this function `get_auth_key_from_xxxx`? 
`verify_xxx` sounds like it should return a boolean value

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 12]
```diff
@@ -518,4 +600,48 @@ module aptos_framework::account {
...
...ab6ee9b5c9d15f4fbcba241362e814650a961daabb3464a19c23cf3b21ddda4b1370d2cf38aebd33732628049c7a00b97d4baf5b221d5ac38ef40f3fc7e8cd5c72229c0427b5fe68f580dbb9e297b544613e6948539e0328f304d79e0796c50344562a4b017019423b3dd6784495bada7e096cb9302f653e9618cf7b063de024db9adfb9d0b77dbb0048d312f167146f2807d991499fe21702f0df26ea83d418ccf8fd5fb6611c0cfb87c01e680e527a0b3fd7c0ac9b6e742db658fda20f93b446ac4b011430f8d455d71e29df2dca88a772bd598d5953f9fab64b9c5f7e0f424cd8659fb033aee099abeab7ce21463a8d4f84b803cd970ad8812196438311a72fc30720f42f863234ad30ba335d91281bf5c2c09e74e4c37b2f8d18781735e1301c0bd24d8b8e564a1e380cd862e4d058b99645efc98f27e1857f38ede528deb50f9310b2ee68756bac74279990de5dfd3557dbb0e631dc83e71e51abfa928bffff19b5e8ec751073fb371e3444272899cef16789a8812c21d54a4ae7fedd80623e2ed80f7652aa6a6ed97813d69c7f636b4a832088ed2d981e809dddf25566b926cb1677ba80793b4aa70087ef27c8b54d4c28f15ad2fe702293e628888fc548851063b89bcd1b0d8547a2856dbd9814bbff331e2e5a6ea7bb9cd6881ea498f46b3552d4815fed985b8c94c7800355f334b0e81fa7e3eb8ed19d83ac19d14d1bf0c4e978f339818514e36e9253172d8cb8ae4b194c2b507cb498e2c15fa567d17ef0be9d8cfd9eb17d46e71c0741bc81def63b6ff5559e83cd0a8f616ffa0b7c0f5b4c874d95ec92e97e7711b85cb2fef5a6f4fd4902";
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:607]: could we create a struct and call `bcs::to_bytes` to generate this bytes?

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 13]
```diff
@@ -518,4 +600,48 @@ module aptos_framework::account {
...
+        let alice_address = from_bcs::to_address(curr_auth_key);
+        let alice = create_account_unchecked(alice_address);
+
+        rotate_authentication_key(&alice, MULTI_ED25519_SCHEME, curr_pk_bytes, MULTI_ED25519_SCHEME, new_pk_bytes, cap_rotate_key, cap_update_table);
+        let address_map = &mut borrow_global_mut<OriginatingAddress>(@aptos_framework).address_map;
+        let new_pk = multi_ed25519::new_unvalidated_public_key_from_bytes(new_pk_bytes);
+        let new_auth_key = multi_ed25519::unvalidated_public_key_to_authentication_key(&new_pk);
+        let new_address = from_bcs::to_address(new_auth_key);
+        let expected_originating_address = table::borrow(address_map, new_address);
+        assert!(*expected_originating_address == alice_address, 0);
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:623]: maybe add an assert on alice's account auth_key updated?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 11 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move]: good point! updated the function name

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 12 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:607]: ahh I think that'd be a bit hard cus I don't know if there's a way to generate ed25519 key / multi_ed25519 keys in move? those bytes were generated in rust

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 13 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:623]: good point - just updated : )

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 14]
```diff
@@ -249,6 +255,82 @@ module aptos_framework::account {
...
+
+        let account_resource = borrow_global_mut<Account>(addr);
+
+        // verify the public key matches the current authentication key
+        if (from_scheme == ED25519_SCHEME) {
+            let from_pk = ed25519::new_unvalidated_public_key_from_bytes(from_public_key_bytes);
+            let from_auth_key = ed25519::unvalidated_public_key_to_authentication_key(&from_pk);
+            assert!(account_resource.authentication_key == from_auth_key, error::unauthenticated(EWRONG_CURRENT_PUBLIC_KEY));
+        } else {
+            let from_pk = multi_ed25519::new_unvalidated_public_key_from_bytes(from_public_key_bytes);
```
Comment (jjleng) [inline aptos-move/framework/aptos-framework/sources/account.move:301]: @chloeqjz @alinush can we also validate that the multisig public keys have at most 32 ed25519 keys in them? It is better to be safe

Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:301]: I think alin already checked this in the rust implementation: https://github.com/aptos-labs/aptos-core/blob/main/crates/aptos-crypto/src/multi_ed25519.rs#L64

Comment (jjleng) [reply aptos-move/framework/aptos-framework/sources/account.move:301]: yes, rust has that. but that restriction doesn't limit users from sending a `to_public_key_bytes ` with more than 32 keys here, right? That means the user might be able to rotate with a multisig key that is invalid and then the account cannot be accessed anymore. Is this the case?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (areshand) [reply aptos-move/framework/aptos-framework/sources/account.move]: why do we need to enforce user have to rotate from account_a? previous code works when user rotate from both account_a and account_b right.

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 14 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:301]: ooh so I think this function 
`verify_key_rotation_signature_and_get_auth_key(to_scheme, to_public_key_bytes, cap_update_table, &challenge)` (https://github.com/aptos-labs/aptos-core/blob/2e9b216d0e184af6d2f8db3d78785fc9697d671f/aptos-move/framework/aptos-framework/sources/account.move#L267) calls `native_signature_verify_strict ` (https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/src/natives/cryptography/multi_ed25519.rs#L58). and `native_signature_verify_strict ` calls `multi_ed25519::MultiEd25519PublicKey::try_from(pubkey.as_slice()) ` and that function calls `check_and_get_threshold` where they have the following code `
    if num_of_keys == 0 || num_of_keys > MAX_NUM_OF_KEYS || threshold_num_of_bytes != 1 {
        Err(CryptoMaterialError::WrongLengthError)
    } ` to check the length. I think you're right that they may be able to send a invalid `to_public_key_bytes` but this should fail in the signature verification step

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move]: I guess we want some kind of standard? and it is more intuitive to rotate it from the original account

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 14 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:301]: Right... you both have good points:
1. @jjleng, we could detect this earlier in `new_unvalidated_public_key_from_bytes`
2. @chloeqjz, we do eventually detect this in the native signature verification.

@chloeqjz, can we nonetheless implement the check you Slack'd me with in `new_unvalidated_public_key_from_bytes`, just to make sure we error as early as possible?
Specifically,
```
assert!(vector::length(&bytes) / INDIVIDUAL_PUBLIC_KEY_NUM_BYTES <= MAX_NUMBER_OF_PUBLIC_KEYS, error::invalid_argument(E_WRONG_PUBKEY_SIZE));
```
and define `MAX_NUMBER_OF_PUBLIC_KEYS` to be 32!

Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:301]: In addition, can you also return false earlier in our `native_public_key_validate` function that implements MultiEd25519 public key validation? Specifically, at line 33 [here](https://github.com/aptos-labs/aptos-core/blob/9366413f678e55f2babd3e7da192d0ecce58ae40/aptos-move/framework/src/natives/cryptography/multi_ed25519.rs#L33), check that the # of sub public keys is not > 32:

```
    if num_sub_pks > multi_ed25519::MAX_NUM_OF_KEYS {
        return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)]));
    }
```

You will need to make MAX_NUM_OF_KEYS public in the MultiEd25519 file from `aptos-crypto/`

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move]: Uhm.... Now, I'm confused: when you rotate to `key_b` you don't create an `account_b` at all. You just update `account_a`'s `authentication_key`. And you update the table, to make `key_b` point to `account_a`'s address.

So what do you mean by "if they call the function from `account_b`", if there's no `account_b`.

@davidiw can you help us here?

Also, @davidiw, what happens if somebody tries to create this non-existent `account_b` **after** the rotation to `key_b` happened?

I think that might cause the problem @chloeqjz is talking about above? Or it might cause other problems...

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 12 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:605]: There isn't: no randomness in Move :'-(

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (areshand) [reply aptos-move/framework/aptos-framework/sources/account.move]: @alinush `I thought when you rotate to key_b, you don't create an account_b at all.` this is no enforcement that the person doesn't have account_b right. In fact if they own the private key to account b, they own account_b in fact.  I think the assumption that they won't or shouldn't call the function from account_b is not valid assumption. make sense?

**Review by areshand — APPROVED** [review]

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs [hunk 15]
```diff
@@ -63,6 +109,88 @@ fn rotate_auth_key_twice() {
...
+            current_account.pubkey.to_bytes().to_vec(),
+            1,
+            new_public_key.to_bytes().to_vec(),
+            signature_by_curr_privkey.to_bytes().to_vec(),
+            signature_by_new_privkey.to_bytes().to_vec(),
+        )
+    ));
+}
+
+pub fn assert_successful_payload_key_rotation_from_ed25519_to_ed25519(
```
Comment (alinush) [inline aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:153]: Maybe not in this PR, but the next one: You can probably use the traits from `aptos-crypto/src/traits.rs` to make this function generic: it is identical to `assert_successful_payload_key_rotation_from_ed25519_to_multi_ed25519` except for the public key types.

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs [hunk 16]
```diff
@@ -63,6 +109,88 @@ fn rotate_auth_key_twice() {
...
+            0,
+            current_account.pubkey.to_bytes().to_vec(),
+            0,
+            new_public_key.to_bytes().to_vec(),
+            signature_by_curr_privkey.to_bytes().to_vec(),
+            signature_by_new_privkey.to_bytes().to_vec(),
+        )
+    ));
+}
+
```
Comment (alinush) [inline aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:193]: Again, perhaps in the next PR, we should also test the following cases:
1. Rotation from MultiEd25519 to Ed25519:
2. Rotation from MultiEd25519 to MultiEd25519: 

Making your functions generic using the `aptos-crypto` traits will help with this.

Comment (chloeqjz) [reply aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:193]: good point - will do that in the next pr : )

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs [hunk 17]
```diff
@@ -30,6 +32,50 @@ struct RotationProofChallenge {
...
+        account2.pubkey.clone(),
+    );
+    // rotate account1's keypair to account2
+    account1.rotate_key(account2.privkey, account2.pubkey);
+    // verify that we can still get to account1's originating address
+    verify_originating_address(&mut harness, account1.auth_key(), *account1.address());
+}
+
+#[test]
+fn rotate_auth_key_ed25519_to_multi_ed25519() {
```
Comment (alinush) [inline aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:57]: Same comment here: In the next PR,  can we coalesce this function and the previous one into one generic function?

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs [hunk 18]
```diff
@@ -30,6 +32,50 @@ struct RotationProofChallenge {
...
+    assert_successful_payload_key_rotation_from_ed25519_to_ed25519(
+        &mut harness,
+        account1.clone(),
+        *account1.address(),
+        10,
+        account2.privkey.clone(),
+        account2.pubkey.clone(),
+    );
+    // rotate account1's keypair to account2
+    account1.rotate_key(account2.privkey, account2.pubkey);
```
Comment (alinush) [inline aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:51]: @chloeqjz I thought the key rotation was called inside `assert_successful_payload_key_rotation_from_ed25519_to_ed25519` [here](https://github.com/aptos-labs/aptos-core/blob/2e9b216d0e184af6d2f8db3d78785fc9697d671f/aptos-move/e2e-move-tests/tests/rotate_auth_key.rs#L181), no?

What does this call do?

You never make this call in`rotate_auth_key_ed25519_to_multi_ed25519`...

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs [hunk 19]
```diff
@@ -30,6 +32,50 @@ struct RotationProofChallenge {
...
+    // assert that the payload is successfully processed (the signatures are correct)
+    assert_successful_payload_key_rotation_from_ed25519_to_multi_ed25519(
+        &mut harness,
+        account1.clone(),
+        *account1.address(),
+        10,
+        private_key,
+        public_key,
+    );
+
```
Comment (alinush) [inline aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:74]: You don't call `rotate_key` here as you do in the other `rotate_auth_key_ed25519_to_ed25519 ` function...

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs (hunk 18 above)
Comment (chloeqjz) [reply aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:51]: the key rotation function was called inside `assert_successful_payload_key_rotation_from_ed25519_to_ed25519` (this function just asserts that the key rotation on-chain is successful), but we're rotating the key in rust / off-chain (updating the keypair to the new keypair, after we successfully updated the auth key on-chain). 

updating the keypair, so that we can verify the address using `account1.auth_key()` (the new one after rotation). otherwise we'll need to pass in `account2.auth_key()` (but `account2` is only for generating a keypair in test, we don't want to create a new account just for key rotation)

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs (hunk 19 above)
Comment (chloeqjz) [reply aptos-move/e2e-move-tests/tests/rotate_auth_key.rs:74]: `rotate_key` only takes in ed25519 public key and private key as parameters (we cannot pass in multied25519 public key and private key in this case, but it's fine. it's a test function). for the other functions, especially the `rotate_auth_key_twice`, we are trying to rotate from the same account twice, so it's necessary to update the private key and public key after the first rotation using `rotate_key`. all the other test functions only rotate once, so we don't really need to do `rotate_key`

File: aptos-move/framework/src/natives/cryptography/multi_ed25519.rs [hunk 20]
```diff
@@ -48,7 +54,7 @@ fn native_public_key_validate(
     }
 
     let num_checked = NumArgs::new(num_checked as u64);
-    let cost = gas_params.base
+    cost = cost
```
Comment (alinush) [inline aptos-move/framework/src/natives/cryptography/multi_ed25519.rs:57]: This will probably fail the lint unless you do `cost +=`

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs (hunk 16 above)
Comment (alinush) [reply aptos-move/e2e-move-tests/tests/rotate_auth_key.rs]: Cool! We'll need to make the `Account` struct generic over any signature scheme [here](https://github.com/aptos-labs/aptos-core/blob/04e39cb28e1b3aef281fbc6993d090d2f7d838e9/aptos-move/e2e-tests/src/account.rs#L37)

File: aptos-move/e2e-move-tests/tests/rotate_auth_key.rs (hunk 18 above)
Comment (alinush) [reply aptos-move/e2e-move-tests/tests/rotate_auth_key.rs]: Quick summary after in-person discussion: the `account1.rotate_key(account2.privkey, account2.pubkey);` was artificial and unnecessary: we instead passed `account2.auth_key()` as an argument to `verify_originating_address`

**Review by alinush — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move]: After clarifying in person/over Slack, here’s the problematic scenario that this PR fixes via [the assert here](https://github.com/aptos-labs/aptos-core/pull/3626/files#diff-ef8a70cd4b03dc7b8b6dfd8306886d480c155211d8c472b59585471e4c60c427R283) 🧯:
1. User rotated `acc_a`’s `key_a` to `key_b`.
2. User creates an account `acc_b` with `key_b` as its authentication key (which is mapped to `acc_a` via the originating address table)
3. User calls rotate_key from `acc_b` (rather than `acc_a`) asking to rotate `acc_b`’s `key_b` to `key_c`

This is where there is a bug. Specifically, our old `rotate_authentication_key_ed25519` implementation (as well as our initial implementation of `rotate_authentication_key` in the current PR) does the following:
1. Fetches `addr_a = OriginatingAddress[key_b]` (See [this change](https://github.com/aptos-labs/aptos-core/pull/3483/files#diff-ef8a70cd4b03dc7b8b6dfd8306886d480c155211d8c472b59585471e4c60c427R238).)
2. Removes `OriginatingAddress[key_b]` (See [the same change](https://github.com/aptos-labs/aptos-core/pull/3483/files#diff-ef8a70cd4b03dc7b8b6dfd8306886d480c155211d8c472b59585471e4c60c427R238).)
3. Sets `OriginatingAddress[key_c] = addr_a`
4. Sets `acc_b`’s (not `acc_a`’s) authentication key to `key_c`. (See [this line.](https://github.com/aptos-labs/aptos-core/pull/3483/files#diff-ef8a70cd4b03dc7b8b6dfd8306886d480c155211d8c472b59585471e4c60c427R249))

As a result, we’ve left the system in an inconsistent state:
1. `acc_a` still has `key_b` as its auth key, but no originating-address mapping
2. `acc_b` has `key_c` as its auth key, with an originating-address mapping it back to `addr_a`.

In other words, the originating address table and the resource at `acc_a` don’t agree on what the current auth key for `acc_a` is.

One way to deal with this is to prevent key rotation from `acc_b` via the assert Chloe just introduced [here](https://github.com/aptos-labs/aptos-core/pull/3626/files#diff-ef8a70cd4b03dc7b8b6dfd8306886d480c155211d8c472b59585471e4c60c427R325). (edited)
